### PR TITLE
Spike plugins instrumentation with Artifact transform

### PR DESCRIPTION
**Pull request by asodja**

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle help

**Comment by asodja** [comment]
@bot-gradle clean

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentArtifactTransform.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        FileHierarchySet globalCacheSet = FileHierarchySet.empty();
+        for (File file : getParameters().getGlobalCacheLocations()) {
+            globalCacheSet = globalCacheSet.plus(file);
+        }
+        if (globalCacheSet.contains(getInputAsFile().getAbsolutePath())) {
+            outputs.file(getInput());
+        } else {
+            File originalFile = outputs.file(getInputAsFile().getName());
+            GFileUtils.copyFile(getInputAsFile(), originalFile);
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentArtifactTransform.java:97]: 🤔 We should not need to copy inputs to outputs. Transforms allow you to emit some of your inputs as outputs as is.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java [hunk 2]
```diff
@@ -27,17 +29,23 @@
...
 import org.gradle.internal.component.local.model.OpaqueComponentIdentifier;
 import org.gradle.internal.logging.util.Log4jBannedVersion;
 import org.gradle.util.GradleVersion;
 
 import java.util.List;
 
 public class DefaultScriptClassPathResolver implements ScriptClassPathResolver {
+
+    private static final Attribute<Boolean> HIERARCHY_COLLECTED_ATTRIBUTE = Attribute.of("hierarchy-collected", Boolean.class);
+    private static final Attribute<Boolean> INSTRUMENTED_ATTRIBUTE = Attribute.of("instrumented", Boolean.class);
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java:48]: We should probably use `org.gradle` prefixes for these to avoid conflicts with attributes that the user may have already set on the plugin variants

**Comment by lptr** [comment]
We need to first solve the problem of fine-grained locking of transforms workspaces.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle help

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle clean
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformBuildOperationIntegrationTest.groovy [hunk 3]
```diff
@@ -1363,12 +1363,12 @@ class ArtifactTransformBuildOperationIntegrationTest extends AbstractIntegration
...
+        getExecutePlannedStepOperations(1)
 
-        buildOperations.progress(IdentifyTransformExecutionProgressDetails).size() == 2
-        buildOperations.all(ExecuteWorkBuildOperationType).size() == 2
-        buildOperations.all(SnapshotTransformInputsBuildOperationType).size() == 2
-        buildOperations.all(ExecuteTransformActionBuildOperationType).size() == 2
+        buildOperations.progress(IdentifyTransformExecutionProgressDetails).size() == 4
+        buildOperations.all(ExecuteWorkBuildOperationType).size() == 4
+        buildOperations.all(SnapshotTransformInputsBuildOperationType).size() == 4
+        buildOperations.all(ExecuteTransformActionBuildOperationType).size() == 4
```
Comment (lptr) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformBuildOperationIntegrationTest.groovy:1371]: 🤔 Why did these increase? Would be nice to have some comments here explaining each count.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy [hunk 4]
```diff
@@ -501,42 +507,54 @@ class BuildScriptClasspathIntegrationSpec extends AbstractIntegrationSpec implem
...
-        inJarCache(filename, false)
+    TestFile notInArtifactTransformCache(String filename) {
+        inArtifactTransformCache(filename, false)
     }
 
-    TestFile inJarCache(String filename, boolean shouldBeFound = true) {
+    TestFile inArtifactTransformCache(String filename, boolean shouldBeFound = true) {
         String fullpath = result.output.readLines().find { it.matches(">>>file:.*${filename}") }.replace(">>>", "")
-        assert fullpath.startsWith(cacheDir.toURI().toString()) == shouldBeFound
+        assert fullpath.startsWith(artifactTransformCacheDir.toURI().toString()) == shouldBeFound
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy:532]: 💭  This re-parsing of console outputs with internal knowledge of cache directory structures is such a hard to follow, brittle way to write these tests. And we have so many of them. We should think about how we could do this better, i.e. without relying on (this much) implementation detail.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java [hunk 5]
```diff
@@ -151,14 +152,21 @@ private void defineConfiguration() {
...
+            dependencyHandler = getAndConfigureDependencyHandler(dependencyResolutionServices);
         }
         if (classpathConfiguration == null) {
             classpathConfiguration = configContainer.migratingUnlocked(CLASSPATH_CONFIGURATION, ConfigurationRolesForMigration.LEGACY_TO_RESOLVABLE_DEPENDENCY_SCOPE);
             buildLogicBuilder.prepareClassPath(classpathConfiguration, dependencyHandler);
         }
     }
 
+    private static DependencyHandler getAndConfigureDependencyHandler(DependencyResolutionServices dependencyResolutionServices) {
+        // TODO: JavaEcosystemSupport.configureSchema is called in the constructor, should we move it here?
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java:164]: ❓ Shall we resolve this TODO?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        File input = getInput().get().getAsFile();
+        if (!input.exists()) {
+            // Don't instrument files that don't exist, these could be files added to classpath via files()
+            return;
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:85]: 🤔 Hah! Did we inherit this? Feels like something we could filter earlier, or disallow.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        ClasspathElementTransform transform = transformFactory.createTransformer(input, new InstrumentingClassTransform(), InstrumentingTypeRegistry.EMPTY);
+        transform.transform(outputFile);
+
+        // Copy original jars after in case they are not in global cache
+        if (instrumentationServices.getGlobalCacheLocations().isInsideGlobalCache(input.getAbsolutePath())) {
+            // The global caches are additive only, so we can use it directly since it shouldn't be deleted or changed during the build.
+            outputs.file(getInput());
+        } else {
+            File copyOfOriginalFile = outputs.file(ORIGINAL_JAR_DIR_NAME + "/" + input.getName());
+            GFileUtils.copyFile(input, copyOfOriginalFile);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:104]: ❓ I asked this before: do we need to copy the original file here? We have also the option to pass `input` directly to `outputs.file()` and then save the time needed to copy the file.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Default implementation of {@link ClasspathBuilder} that is registered also as a service.
+ *
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java:33]: :nail_care: 
```suggestion
 * <p>
```

File: subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ * This implementation first writes Jar file to a temp file and then moves the result to the destination file.
+ * You can use @{@link InPlaceClasspathBuilder} if you want to avoid this indirection and write directly to the destination file.
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java:35]: :thinking: maybe a short rational when the user should prefer one over the other. I suppose, one should use `InPlaceClasspathBuilder` when the atomicity of the output JAR creation is enforced by some other means.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java [hunk 10]
```diff
@@ -64,29 +94,55 @@ public void prepareClassPath(Configuration configuration, DependencyHandler depe
...
+    }
+
+    private static FileCollection getInstrumentedExternalDependencies(Configuration classpathConfiguration) {
+        return classpathConfiguration.getIncoming().artifactView((Action<? super ArtifactView.ViewConfiguration>) config -> {
+            config.attributes(it -> it.attribute(INSTRUMENTED_ATTRIBUTE, INSTRUMENTED_EXTERNAL_DEPENDENCY_ATTRIBUTE));
+            config.componentFilter(componentId -> {
+                if (componentId instanceof OpaqueComponentIdentifier) {
+                    DependencyFactoryInternal.ClassPathNotation classPathNotation = ((OpaqueComponentIdentifier) componentId).getClassPathNotation();
+                    return !DefaultScriptClassPathResolver.NO_GRADLE_API.contains(classPathNotation);
+                }
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java:136]: :nail_care: maybe it is still worth it to have a separate "isGradleApi" method.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java [hunk 11]
```diff
@@ -15,37 +15,67 @@
...
 import org.gradle.util.GradleVersion;
 
+import java.util.EnumSet;
+import java.util.Set;
+
+import static org.gradle.api.artifacts.type.ArtifactTypeDefinition.JAR_TYPE;
+
 public class DefaultScriptClassPathResolver implements ScriptClassPathResolver {
+
+    private static final Set<DependencyFactoryInternal.ClassPathNotation> NO_GRADLE_API = EnumSet.copyOf(ImmutableSet.of(
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java:57]: :question: why is the intermediate `ImmutableSet` needed? You aren't getting immutability, and there's `EnumSet.of` static factory method to build the set out of values.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy [hunk 12]
```diff
@@ -501,42 +507,54 @@ class BuildScriptClasspathIntegrationSpec extends AbstractIntegrationSpec implem
...
-    List<File> getCachedTransformedJarsByName(String jarName) {
-        Arrays.stream(cacheDir.listFiles()).filter {
-            File cacheChild -> isCachedTransformedEntryDir(cacheChild)
-        }.map {
-            File cacheChild -> new File(cacheChild, jarName)
-        }.filter {
-            it.exists()
-        }.collect(Collectors.toList())
+    List<File> getArtifactTransformJarsByName(String jarName) {
+        return Files.find(artifactTransformCacheDir.toPath(), 4, (path, attributes) -> normaliseFileSeparators(path.toString()).endsWith(jarName))
```
Comment (mlopatkin) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy:555]: :nail_care: could you please extract 4 into a variable to save a trip to javadoc? maybe with a comment why this depth is enough.

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 13]
```diff
@@ -236,6 +246,99 @@ public String toString() {
...
+        List<File> inputFiles = classPath.getAsFiles();
+
+        Builder result = builderWithExactSize(resultSize);
+        for (int i = 0; i < inputFiles.size(); ++i) {
+            File inputFile = inputFiles.get(i);
+            if (isInstrumentedMarkerFile(inputFile)) {
+                checkArgument(i + 2 < inputFiles.size(), "Missing the instrumented or original JAR for classpath %s", inputFiles);
+                File instrumentedJar = inputFiles.get(i + 1);
+                File originalJar = inputFiles.get(i + 2);
+                checkArgument(areInstrumentedAndOriginalJarValid(instrumentedJar, originalJar), "Instrumented JAR doesn't match original JAR %s", originalJar.getName());
```
Comment (mlopatkin) [inline platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:320]: :nail_care: we should probably also add the instrumented jar name too. And I think printing full paths would be better, because e.g. parent dirs participate in the check too.
```suggestion
                checkArgument(areInstrumentedAndOriginalJarValid(instrumentedJar, originalJar), "Instrumented JAR %s doesn't match original JAR %s", instrumentedJar.getAbsolutePath(), originalJar.getAbsolutePath());
```

**Review by mlopatkin — COMMENTED** [review]
:+1: Looks good overall, I left a few minor suggestions.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java (hunk 8 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java]: Fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java (hunk 9 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultClasspathBuilder.java]: Improved the comment

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java (hunk 10 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java]: Makes sense, done

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java (hunk 11 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java]: Good catch, fixed

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy (hunk 12 above)
Comment (asodja) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy]: I think we don't have any problem with any depth here, I changed it to Integer.MAX_VALUE

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java (hunk 13 above)
Comment (asodja) [reply platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java]: Makes sense

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy (hunk 4 above)
Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathIntegrationSpec.groovy:532]: Let's add an issue for this.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java (hunk 6 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:77]: Is there something we can capture of that learning in a comment here for the next person that tries to remove this?

Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:77]: Improved the comment

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
> Though `inferParameterType` looks scary

It should work fine, worst case we can revert that part if there will be any issue.
